pkg/reconcile/pipeline/context/impl.go

```diff
@@ -118,15 +121,44 @@ func (i *impl) Services() ([]pipeline.Service, error) {
 func (i *impl) Applications() ([]pipeline.Application, error) {
 	if i.applications == nil {
 		if ref := i.serviceBinding.Spec.Application; ref != nil {
-			gvr, err := i.typeLookup.ResourceForReferable(ref)
-			if err != nil {
-				return nil, err
+			if i.serviceBinding.Spec.Application.Name != "" &&
+				i.serviceBinding.Spec.Application.LabelSelector != nil &&
+				i.serviceBinding.Spec.Application.LabelSelector.MatchLabels != nil {
+				return nil, errs.New("Both label selector and name for application should be used")
 			}
-			u, err := i.client.Resource(*gvr).Namespace(i.serviceBinding.Namespace).Get(context.Background(), ref.Name, metav1.GetOptions{})
-			if err != nil {
-				return nil, err
+			if i.serviceBinding.Spec.Application.Name != "" {
+				gvr, err := i.typeLookup.ResourceForReferable(ref)
+				if err != nil {
+					return nil, err
+				}
+				u, err := i.client.Resource(*gvr).Namespace(i.serviceBinding.Namespace).Get(context.Background(), ref.Name, metav1.GetOptions{})
+				if err != nil {
+					return nil, err
+				}
+				i.applications = append(i.applications, &application{gvr: gvr, persistedResource: u, bindingPath: i.serviceBinding.Spec.Application.BindingPath})
+
+			}
+			if i.serviceBinding.Spec.Application.LabelSelector != nil && i.serviceBinding.Spec.Application.LabelSelector.MatchLabels != nil {
+				gvr, err := i.typeLookup.ResourceForReferable(ref)
```

Comment:
this lines repeats also in line 130 - so it could be moved before back to the original place.

Reply:
Done. 5326c6f674cb869fdaa8242bd4cc454fb72941cc